### DISCUSSION TITLE
Multi-threaded RPC client

### DISCUSSION THREAD

**Discussion by CarlSayres** [activej/activej#255] [Q&A]
I have a multi-threaded RPC server working. (Thanks for the previous help with this.)

But now I need to also create a multi-threaded RPC client. So far, I've made this work by creating an GenericObjectPool of launcher objects, with each launcher using it's own thread.  Whenever a thread need to make an RPC call, it borrows a launcher from the pool, make the call, then returns it to the pool. This does work, but it's thread-heavy.

Is there a way to do this using a Worker EventLoop?  Would this be more efficient?

Thanks,
Carl

**Comment by eduard-vasinskyi** [comment]
Hi, @CarlSayres

ActiveJ’s Launcher is meant to be launched as a standalone application. 

If you need multiple worker RpcClients, you can once again use `WorkerPoolModule`. Just install the `WorkerPoolModule` in your Launcher and provide a `WorkerPool` of the required size. Then, provide a worker Eventloop and RpcClient (using `@Worker` annotations). After that, you can request `WorkerPool.Instances<RpcClient>` from DI. `WorkerPool.Instances` is an iterable of worker instances. You can iterate through RPC clients and send RPC requests using the required worker instance. 

Just make sure that the request is sent from within the correct worker Eventloop. You can call `RpcClient#getEventloop` in order to obtain a worker Eventloop that corresponds to a worker RpcClient.
